testkit/src/main/java/org/apache/calcite/test/SqlOperatorTest.java

```diff
@@ -6215,6 +6215,71 @@ void checkRegexpExtract(SqlOperatorFixture f0, FunctionAlias functionAlias) {
     f.checkNull("log(10, cast(null as real))");
   }
 
+  /** Test case for
+   * <a href="https://issues.apache.org/jira/browse/CALCITE-6224">[CALCITE-6224]
+   * Add LOG@ function (enabled in MYSQL, Spark library)</a>. */
+  @Test void testLog2Func() {
+    final SqlOperatorFixture f0 = fixture();
+    final Consumer<SqlOperatorFixture> consumer = f -> {
+      f.setFor(SqlLibraryOperators.LOG2);
+      f.checkScalarApprox("log2(2)", "DOUBLE NOT NULL",
+          isWithin(1.0, 0.000001));
+      f.checkScalarApprox("log2(4)", "DOUBLE NOT NULL",
+          isWithin(2.0, 0.000001));
+      f.checkScalarApprox("log2(65536)", "DOUBLE NOT NULL",
+          isWithin(16.0, 0.000001));
+      f.checkScalarApprox("log2(-2)", "DOUBLE NOT NULL",
+          "NaN");
+      f.checkScalarApprox("log2(2/3)", "DOUBLE NOT NULL",
+          "-Infinity");
+      f.checkScalarApprox("log2(2.2)", "DOUBLE NOT NULL",
+          "1.1375035237499351");
+      f.checkScalarApprox("log2(0.5)", "DOUBLE NOT NULL",
+          "-1.0");
+      f.checkScalarApprox("log2(3)", "DOUBLE NOT NULL",
+          isWithin(1.5849625007211563, 0.000001));
+      f.checkNull("log2(cast(null as real))");
+    };
+    f0.forEachLibrary(list(SqlLibrary.MYSQL, SqlLibrary.SPARK), consumer);
+  }
+
+  /** Test case for
+   * <a href="https://issues.apache.org/jira/browse/CALCITE-6232">[CALCITE-6232]
+   * Using fractions in LOG function does not return correct results</a>. */
+  @Test void testLogFuncByConvert() {
+    final SqlOperatorFixture f0 = fixture();
+    f0.setFor(SqlLibraryOperators.LOG, VmName.EXPAND);
+    final SqlOperatorFixture f = f0.withLibrary(SqlLibrary.BIG_QUERY);
+    f.checkScalarApprox("log(2.0/3, 2)", "DOUBLE NOT NULL",
```

Comment:
Hello, I seem to have forgotten the simplest truth, that is, there are no fractions in computers, only floating point numbers. I changed 2/3 in the test to 2.0/3, and it turns out that the result is correct. If 2/3 must output the correct result, then this bug does not only exist in the log function, but the pow and abs functions also have similar problems. Regarding log(0) I don't think it's a bug.

Thank you very much for your guidance on these two PRs of my adaptation function. I am sorry for causing you a lot of trouble. If there is anything missing in this PR, please point it out, thank you.

Reply:
I will remove this part of the test later

Reply:
I would recommend studying the SQL language, otherwise you will have surprises.
It's very difficult to contribute to a compiler if you don't understand the language you are compiling.
In SQL floating point is generally not used. SQL has a family of DECIMAL (or NUMERIC -- same thing) data types, which are very different from floating point. 2.0 is a DECIMAL literal. DECIMAL is not a type, it's a family of types with different scales and precisions. Unfortunately, different dialects of SQL treat DECIMAL in incompatible ways. Floating point is an IEEE standard, but DECIMAL isn't.

Reply:
I am still of the belief that even though LOG(0) returning `-Infinity` makes sense mathematically, we should do our best to parity the dialects that support these functions. I am a fan of extending a function's behavior, but not augmenting it.

Reply:
If a function is part of a dialect, it should definitely have the same behavior as the dialect requires. This should also be documented. We may be dealing with multiple LOG functions, with slightly different behaviors. That's why it is very important to test all corner cases, including writing negative tests.

Reply:
Totally agree, if necessary we can do something similar to the `SUBSTR` functions which I think is implemented per-dialect.

Reply:
I have a question. For example, the log2(0) function has different expressions in the results of spark and mysql. Do we need to adapt them separately?
Regarding negative tests, I would like to ask which ones are negative tests. At first I thought it was just log2(0).

Reply:
I have tested in both spark and mysql. The output of log2(0) in spark and mysql is null. Do I need to process the parameter 0 in sqlfunctions? Or simply adding test is enough.

Reply:
Hello, thank you very much for your suggestion. Regarding negative test, I think of the following case

- log10(-1): When the parameter in log10 is a negative number
- 
- log10(null): When the parameters in log10 are empty
- 
- log10(-`Infinity`): When the parameter in log10 is -`Infinity`
- 
- log10(0): When the parameter in log10 is 0

I tested based on mysql. Only when log(0) and log10(-1), calcite cannot match the results in mysql. In mysql, both of them return NULL. When log10(0), calcite returns -`Infinity. `log10(-1) returns NaN

![test2](https://github.com/apache/calcite/assets/77189278/c29a69b6-8c4e-4d94-b7c3-148da99f3243)

Reply:
My current idea is to add a judgment to the log function in sqlFunctions. If it is less than or equal to 0, it will return 0. But there is a problem, that is, log10 and log functions are adapted to the current calcite database. So what should we do? To solve this problem, we cannot guarantee that the results returned by the database will be the same.
In other words, I should write Sqlfunctions specifically for log2

Reply:
@mihaibudiu Or should we deal with this situation?

Reply:
Double.NaN == null？

Reply:
https://github.com/apache/calcite/pull/3669, It seems that there should be the same situation

Reply:
Yes, you need to adjust/add behavior s.t. when this function is used for these dialects, it follows the proper behavior.

Reply:
The examples you gave above are **not** negative tests (aside from the `null` one). Please read my previous comment. What do you expect of the user? It is a numeric function that is designed to support numeric arguments. Anything outside of that is unexpected.

Reply:
I probably understand what you mean. What you mean is that if you input types such as string and boolean that are not numbers, an error should be reported.

Reply:
Since my pr has a conflict, I will close it first, and then I will raise a pr after solving the problem of log2(0) returning null.
Thanks